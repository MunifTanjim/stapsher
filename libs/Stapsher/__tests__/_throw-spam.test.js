const _ = require('lodash')

const { loadConfig } = require('../../../configs/client')

const helpers = require('../../../__tests__/helpers')

helpers.addSnapshotSerializers()

const akismet = require('../../Akismet')
akismet.akismetCheckSpam = jest.fn()

const Stapsher = require('../../Stapsher')

const extraInfo = {
  clientIP: '127.0.0.1',
  clientUserAgent: 'TheMachine/1.0',
  clientReferrer: 'r007'
}

let mockConfig = {}

describe('libs/Stapsher:__throwSpam', () => {
  const parameters = helpers.getParameters()

  let stapsher
  beforeEach(() => {
    akismet.akismetCheckSpam.mockClear()

    mockConfig = helpers.readConfigFile()[parameters.entryType]
    mockConfig = loadConfig(mockConfig).getProperties()
    mockConfig.akismet = {
      ...mockConfig.akismet,
      enable: true,
      apiKey: 'zxcvb',
      siteUrl: 'https://example.com'
    }

    stapsher = new Stapsher(parameters)

    stapsher.config = {
      get: key => _.get(mockConfig, key)
    }

    stapsher.rawFields = helpers.getFields()

    stapsher.addExtraInfo(extraInfo)
  })

  it('returns early if akismet disabled', () => {
    mockConfig.akismet.enable = false

    stapsher.__throwSpam()

    expect(akismet.akismetCheckSpam).toHaveBeenCalledTimes(0)
  })

  it('calls necessary functions', () => {
    stapsher.__throwSpam()

    expect(akismet.akismetCheckSpam).toHaveBeenCalledTimes(1)
    expect(akismet.akismetCheckSpam.mock.calls[0]).toMatchSnapshot()
  })

  it('returns nothig if spam is false', () => {
    akismet.akismetCheckSpam.mockResolvedValueOnce(false)

    expect(stapsher.__throwSpam()).resolves.toBeUndefined()
  })

  it('throws error if spam is true', () => {
    akismet.akismetCheckSpam.mockResolvedValueOnce(true)

    expect(stapsher.__throwSpam()).rejects.toMatchSnapshot()
  })
})
