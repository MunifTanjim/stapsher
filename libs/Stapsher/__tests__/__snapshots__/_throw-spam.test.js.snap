// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`libs/Stapsher:__throwSpam calls necessary functions 1`] = `
[
  "zxcvb",
  "https://example.com",
  {
    "user_ip": "127.0.0.1",
    "user_agent": "TheMachine/1.0",
    "referrer": "r007",
    "permalink": "",
    "comment_type": "comment",
    "comment_author": "Samantha Groves",
    "comment_author_email": "samantha@example.com",
    "comment_author_url": "https://samantha.example.com",
    "comment_content": "We might as well be a symphony"
  }
]
`;

exports[`libs/Stapsher:__throwSpam throws error if spam is true 1`] = `
{
  "code": "AKISMET_IS_SPAM",
  "statusCode": 400,
  "cause": {
    "entryObject": {
      "user_ip": "127.0.0.1",
      "user_agent": "TheMachine/1.0",
      "referrer": "r007",
      "permalink": "",
      "comment_type": "comment",
      "comment_author": "Samantha Groves",
      "comment_author_email": "samantha@example.com",
      "comment_author_url": "https://samantha.example.com",
      "comment_content": "We might as well be a symphony"
    }
  }
}
`;
